File: genkit-tools/common/src/types/agent.ts
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { z } from 'zod';
import { RuntimeErrorSchema } from './error';
import { MessageSchema, ModelResponseChunkSchema } from './model';
import {
  PartSchema,
  ToolRequestPartSchema,
  ToolResponsePartSchema,
} from './parts';

/**
 * Zod schema for an artifact produced during a session.
 */
export const ArtifactSchema = z.object({
  /** Name identifies the artifact (e.g., "generated_code.go", "diagram.png"). */
  name: z.string().optional(),
  /** Parts contains the artifact content (text, media, etc.). */
  parts: z.array(PartSchema),
  /** Metadata contains additional artifact-specific data. */
  metadata: z.record(z.any()).optional(),
});
export type Artifact = z.infer<typeof ArtifactSchema>;

/**
 * Zod schema for a snapshot's lifecycle status.
 *
 * - `pending`: a detached invocation is still processing the queued inputs.
 *   The snapshot's state is empty until the background work finishes, at
 *   which point it is rewritten with the cumulative final state and a
 *   terminal status.
 * - `aborting`: the `abort` companion action stopped a detached invocation
 *   and its worker is winding down toward the write that stamps the state
 *   on the row. Not terminal and not yet resumable: the worker keeps
 *   refreshing `heartbeatAt` while it drains, and the row settles as
 *   `aborted`. Written by runtimes whose abort stops the work before the
 *   state is recorded; a runtime that records both in one write goes to
 *   `aborted` directly.
 * - `completed`: the snapshot captures a settled state.
 * - `aborted`: the snapshot's invocation was aborted via the
 *   `abort` companion action while detached.
 * - `failed`: a turn ended with an error. The snapshot's `error` field
 *   describes the failure, and its state is what the turn committed before
 *   it: the conversation trimmed back to the last completed tool round.
 *   Resume is permitted, so whether the failure is worth another attempt is
 *   the client's decision, taken from `error.status`.
 * - `expired`: a `pending` or `aborting` snapshot whose detached background
 *   worker is presumed dead because its heartbeat went stale. Computed on
 *   read from a stale `heartbeatAt`; never persisted (the dead worker can no
 *   longer write a terminal status itself).
 */
export const SnapshotStatusSchema = z.enum([
  'pending',
  'aborting',
  'completed',
  'aborted',
  'failed',
  'expired',
]);
export type SnapshotStatus = z.infer<typeof SnapshotStatusSchema>;

/**
 * Zod schema for the reason an agent turn or invocation finished.
 *
 * The first group mirrors the model-level `FinishReasonSchema` so a turn
 * backed by a single `generate` call can forward its reason verbatim:
 *
 * - `stop`: the model stopped naturally.
 * - `length`: generation hit the token limit.
 * - `blocked`: generation was blocked (e.g. safety).
 * - `interrupted`: the model paused on a tool request awaiting input
 *   (e.g. human approval); the turn can be resumed with a `resume` payload.
 * - `other` / `unknown`: anything else / unspecified.
 *
 * The remaining values are agent-specific outcomes that never arise from
 * forwarding a model finish reason (the model-level namesakes of `aborted`
 * and `failed` accompany generate errors, which the agent reports as a
 * failed turn instead):
 *
 * - `aborted`: the turn or invocation was aborted (e.g. a detached
 *   invocation aborted via the `abort` companion action).
 * - `detached`: the invocation was moved to the background (the client
 *   detached). The returned output reports `detached`; the persisted
 *   snapshot is later finalized with how the background work actually ended.
 * - `failed`: the turn or invocation terminated with an error.
 */
export const AgentFinishReasonSchema = z.enum([
  'stop',
  'length',
  'blocked',
  'interrupted',
  'other',
  'unknown',
  'aborted',
  'detached',
  'failed',
]);
export type AgentFinishReason = z.infer<typeof AgentFinishReasonSchema>;

/**
 * Zod schema for session state.
 */
export const SessionStateSchema = z.object({
  /**
   * ID of the session (conversation) this state belongs to.
   * Framework-owned: assigned when the conversation's first invocation
   * starts and re-stamped on outbound state, so client-managed callers
   * can round-trip the state object opaquely without tracking a separate
   * identifier. For server-managed agents the snapshot row's `sessionId`
   * is canonical and this field mirrors it.
   */
  sessionId: z.string().optional(),
  /** Conversation history (user/model exchanges). */
  messages: z.array(MessageSchema).optional(),
  /** User-defined state associated with this conversation. */
  custom: z.any().optional(),
  /** Named collections of parts produced during the conversation. */
  artifacts: z.array(ArtifactSchema).optional(),
});
export type SessionState = z.infer<typeof SessionStateSchema>;

/**
 * Zod schema for agent input (per-turn).
 *
 * An input with neither `message` nor `resume` continues the conversation
 * from where it stands, which is how a failed turn is re-attempted: resume
 * the `failed` snapshot, send an empty input, and the agent runs the turn
 * again on the messages it committed. It is rejected on a conversation with
 * no messages to continue.
 */
export const AgentInputSchema = z.object({
  /**
   * Detach signals that the client wishes to disconnect after this input is
   * accepted. The server writes a single pending snapshot (with empty
   * state), returns AgentOutput with that snapshot ID, and continues
   * processing any already-buffered inputs in a background context.
   * Streamed chunks emitted after detach are not forwarded over the wire;
   * only the final cumulative state is captured when the snapshot is
   * finalized (or the snapshot is aborted via `abort`).
   */
  detach: z.boolean().optional(),
  /** User's input message for this turn. */
  message: MessageSchema.optional(),
  /** Options for resuming an interrupted generation. */
  resume: z
    .object({
      respond: z.array(ToolResponsePartSchema).optional(),
      restart: z.array(ToolRequestPartSchema).optional(),
    })
    .optional(),
});
export type AgentInput = z.infer<typeof AgentInputSchema>;

/**
 * Zod schema for agent initialization.
 */
export const AgentInitSchema = z.object({
  /**
   * Identifies the session (conversation) to resume or start. Only valid
   * when the agent is server-managed (a session store is configured);
   * mutually exclusive with state (a client-managed conversation carries
   * its identity inside `state.sessionId`). Alone it resumes the session
   * from its latest snapshot: the most recently updated row, whatever its
   * status. If that row is an aborted or still-pending dead end the resume
   * is rejected (pass snapshotId to continue from a specific earlier
   * point); a `failed` row resumes with what the failed turn committed; if
   * the session's history was forked by re-resuming an earlier snapshot,
   * the most recently updated branch wins. If the session has no snapshots
   * yet, a brand-new conversation is started under this caller-chosen ID,
   * and every snapshot it persists carries it. Combined with snapshotId, it
   * asserts which session the snapshot belongs to, and a mismatch is
   * rejected.
   */
  sessionId: z.string().optional(),
  /**
   * Loads state from a persisted snapshot (server-managed state only).
   * May be combined with sessionId to validate that the snapshot belongs
   * to that session. Mutually exclusive with state.
   */
  snapshotId: z.string().optional(),
  /**
   * Direct state for the invocation (client-managed state only). The
   * conversation's identity rides inside it (`state.sessionId`): the
   * framework mints one on the conversation's first invocation and
   * echoes it on the output state, so resending the state object keeps
   * the identity without tracking a separate field. Mutually exclusive
   * with sessionId and snapshotId.
   */
  state: SessionStateSchema.optional(),
});
export type AgentInit = z.infer<typeof AgentInitSchema>;

/**
 * Zod schema for agent result.
 */
export const AgentResultSchema = z.object({
  /** Last model response message from the conversation. */
  message: MessageSchema.optional(),
  /** Artifacts produced during the session. */
  artifacts: z.array(ArtifactSchema).optional(),
  /**
   * Why the invocation finished. Set by a custom agent to override the
   * default (the last turn's reason); omitted to accept the default.
   */
  finishReason: AgentFinishReasonSchema.optional(),
});
export type AgentResult = z.infer<typeof AgentResultSchema>;

/**
 * Zod schema for agent output.
 */
export const AgentOutputSchema = z.object({
  /**
   * ID of the session this invocation belongs to, assigned by the
   * framework when the invocation starts. With server-managed state, a
   * fresh invocation adopts the caller-supplied session ID (see
   * AgentInit.sessionId) or mints a new one, resumed invocations inherit
   * the chain's, and resuming a snapshot from before session IDs existed
   * mints a fresh one. With client-managed state it echoes the ID
   * carried inside the state object (`state.sessionId`), minting one on
   * the conversation's first invocation; only a session with persisted
   * snapshots can be resumed by this ID.
   */
  sessionId: z.string().optional(),
  /**
   * ID of the most recent turn-end snapshot for this invocation. Empty
   * when no store is configured or no turn committed. When `finishReason`
   * is `detached` it is the pending detach snapshot. When `failed`, it is
   * the resume point: the failed turn's own snapshot when the turn
   * committed anything, otherwise the last committed turn's snapshot.
   */
  snapshotId: z.string().optional(),
  /**
   * Final conversation state (only when client-managed). When
   * `finishReason` is `failed`, this is the resume point: what the failed
   * turn committed, or the last successful turn's state when the turn
   * failed before committing anything.
   */
  state: SessionStateSchema.optional(),
  /** Last model response message from the conversation. */
  message: MessageSchema.optional(),
  /** Artifacts produced during the session. */
  artifacts: z.array(ArtifactSchema).optional(),
  /**
   * Why the invocation finished. `detached` when the client detached and
   * the work continues in the background; `failed` when the invocation
   * ended in failure (see `error`); otherwise the last turn's reason
   * (or the value a custom agent set on its result).
   */
  finishReason: AgentFinishReasonSchema.optional(),
  /**
   * Structured failure information when the invocation ended in failure
   * (`finishReason` is `failed`). `status` preserves the original error
   * category (e.g. `INVALID_ARGUMENT`, `FAILED_PRECONDITION`,
   * `INTERNAL`) so callers can still branch on it.
   */
  error: RuntimeErrorSchema.optional(),
});
export type AgentOutput = z.infer<typeof AgentOutputSchema>;

/**
 * Zod schema for the turn-end signal emitted by an agent.
 *
 * A TurnEnd value is emitted exactly once per turn, regardless of whether a
 * snapshot was persisted. Grouping all turn-end signals here lets callers
 * detect turn boundaries with a single field check and leaves room for
 * additional turn-end metadata in the future.
 */
export const TurnEndSchema = z.object({
  /**
   * ID of the snapshot persisted at the end of this turn, whether it
   * succeeded or failed. Empty if no snapshot was written (no store
   * configured, a turn that failed before committing anything, or
   * snapshots were suspended after detach).
   */
  snapshotId: z.string().optional(),
  /**
   * Why this turn finished (e.g. `stop`, `length`, `interrupted`). Lets a
   * caller react to a turn boundary (e.g. pause on `interrupted`) without
   * scanning the message content. Omitted when the turn reported no reason.
   *
   * `failed` reports a failed turn; unless the agent recovers and keeps
   * processing, the invocation then resolves with a failed output carrying
   * the error and the last-good state.
   */
  finishReason: AgentFinishReasonSchema.optional(),
});
export type TurnEnd = z.infer<typeof TurnEndSchema>;

/**
 * Zod schema for the operation kind of a JSON Patch operation (RFC 6902).
 */
export const JsonPatchOpSchema = z.enum([
  'add',
  'remove',
  'replace',
  'move',
  'copy',
  'test',
]);
export type JsonPatchOp = z.infer<typeof JsonPatchOpSchema>;

/**
 * Zod schema for a single RFC 6902 (JSON Patch) operation.
 */
export const JsonPatchOperationSchema = z.object({
  op: JsonPatchOpSchema,
  /** A JSON Pointer (RFC 6901) to the target location, e.g. `"/agentStatus"`. */
  path: z.string(),
  /** Source pointer; required for `move` and `copy`. */
  from: z.string().optional(),
  /** New value; required for `add`, `replace`, and `test`. */
  value: z.any().optional(),
});
export type JsonPatchOperation = z.infer<typeof JsonPatchOperationSchema>;

/**
 * Zod schema for an RFC 6902 JSON Patch: an ordered list of operations
 * applied in sequence.
 */
export const JsonPatchSchema = z.array(JsonPatchOperationSchema);
export type JsonPatch = z.infer<typeof JsonPatchSchema>;

/**
 * Zod schema for agent stream chunk.
 */
export const AgentStreamChunkSchema = z.object({
  /** Generation tokens from the model. */
  modelChunk: ModelResponseChunkSchema.optional(),
  /**
   * An RFC 6902 JSON Patch describing a delta applied to the session's custom
   * state. Emitted automatically whenever the agent mutates custom state, so
   * the client can apply it to its tracked copy and keep custom live as the
   * turn streams. Pointers are rooted at the custom document (e.g.
   * `/agentStatus`), with no `/custom` prefix. The first patch of every turn is
   * a whole-document replace at the root pointer (`""`) that re-bases clients
   * which may not share the server's baseline; subsequent patches are
   * incremental diffs against the last sent value.
   */
  customPatch: JsonPatchSchema.optional(),
  /** A newly produced artifact. */
  artifact: ArtifactSchema.optional(),
  /**
   * Non-null when the agent has finished processing the current input.
   * Groups all turn-end signals; the client should stop iterating and may
   * send the next input.
   */
  turnEnd: TurnEndSchema.optional(),
});
export type AgentStreamChunk = z.infer<typeof AgentStreamChunkSchema>;

/**
 * Zod schema for a persisted point-in-time capture of session state. It is
 * the canonical record written to and read from a session store; the wire
 * representation is shared across language runtimes and the Dev UI.
 */
export const SessionSnapshotSchema = z.object({
  /** Unique identifier for this snapshot (UUID). */
  snapshotId: z.string(),
  /**
   * ID of the session this snapshot belongs to. Assigned by the agent
   * framework when the conversation's first invocation starts and
   * stamped on every later snapshot in the chain, including across
   * resumed invocations. Stores preserve it across rewrites; rows
   * written without one (data from before session IDs existed) belong
   * to no session.
   */
  sessionId: z.string().optional(),
  /**
   * ID of the previous snapshot in this timeline. Informational lineage
   * (for debugging and UI history trees); plays no part in resolving a
   * session's latest snapshot.
   */
  parentId: z.string().optional(),
  /** When the snapshot was first written (RFC 3339). */
  createdAt: z.string(),
  /**
   * When the snapshot's state was last written (RFC 3339). Equals `createdAt`
   * until rewritten; a heartbeat refresh on a pending snapshot does not advance
   * it, so liveness stays distinct from state changes.
   */
  updatedAt: z.string().optional(),
  /**
   * Heartbeat timestamp (RFC 3339) refreshed periodically while a detached
   * (background) turn is in flight. Used to detect a dead background worker:
   * if a `pending` snapshot's heartbeat goes stale (older than the configured
   * timeout), reads surface its status as `expired` (the dead worker can no
   * longer persist a terminal status itself).
   */
  heartbeatAt: z.string().optional(),
  /** Lifecycle state of this snapshot. Empty is treated as `completed`. */
  status: SnapshotStatusSchema.optional(),
  /**
   * Semantic reason the turn or invocation captured here ended (e.g.
   * `stop`, `interrupted`, `failed`, `aborted`). Complements `status` (the
   * persistence lifecycle) so a resumed or background task can report how it
   * ended without re-deriving it from the messages.
   */
  finishReason: AgentFinishReasonSchema.optional(),
  /** Structured failure information for a snapshot in `failed` status. */
  error: RuntimeErrorSchema.optional(),
  /**
   * Conversation state captured at this point. Empty on a pending snapshot
   * (the live state is not yet committed); populated on terminal snapshots
   * with the cumulative final state.
   */
  state: SessionStateSchema.optional(),
});
export type SessionSnapshot = z.infer<typeof SessionSnapshotSchema>;

/**
 * Zod schema for the input of an agent's `getSnapshot` companion action.
 * The action is registered under the agent's name (action type
 * `agent-snapshot`) when the agent has a session store configured. The
 * action returns the stored `SessionSnapshot`, with any configured state
 * transform applied to its state.
 *
 * At least one of `snapshotId` or `sessionId` must be set; they are not
 * mutually exclusive. `snapshotId` fetches a specific snapshot;
 * `sessionId` alone fetches the session's latest snapshot (via the
 * store's `GetLatestSnapshot`, whatever its status). When both are set
 * the fetched snapshot must belong to that session, or the request is
 * rejected.
 */
export const GetSnapshotRequestSchema = z.object({
  /**
   * Identifies the snapshot to fetch. Optional when `sessionId` is given;
   * when both are present the fetched snapshot must belong to that session.
   */
  snapshotId: z.string().optional(),
  /**
   * Identifies the session whose latest snapshot to fetch. Optional when
   * `snapshotId` is given. The latest snapshot is the session's most
   * recently updated row regardless of status (pending, failed, or
   * aborted included).
   */
  sessionId: z.string().optional(),
  /**
   * Returns the snapshot's metadata only: status, finish reason, parent,
   * session, timestamps, and error, with no state payload. The read is
   * shaped exactly as a full read (status defaulting and heartbeat expiry
   * need only the metadata), and a store that can read a row without
   * loading its state does so. For callers that dispatch on where a task
   * stands, this skips loading and serializing a potentially large
   * conversation history.
   */
  metadataOnly: z.boolean().optional(),
});
export type GetSnapshotRequest = z.infer<typeof GetSnapshotRequestSchema>;

/**
 * Zod schema for the input of the `abort` companion action.
 */
export const AgentAbortRequestSchema = z.object({
  /** Identifies the snapshot whose invocation should be aborted. */
  snapshotId: z.string(),
});
export type AgentAbortRequest = z.infer<typeof AgentAbortRequestSchema>;

/**
 * Zod schema for the output of the `abort` companion action.
 */
export const AgentAbortResponseSchema = z.object({
  /** Identifies the snapshot the abort attempt targeted. */
  snapshotId: z.string(),
  /**
   * Snapshot's status after the abort attempt. For a pending snapshot this
   * is `aborting` when the runtime stops the work before it records the
   * state, or `aborted` when it records both in one write. For an
   * already-terminal snapshot this is the existing terminal status (the
   * abort is a no-op).
   */
  status: SnapshotStatusSchema.optional(),
});
export type AgentAbortResponse = z.infer<typeof AgentAbortResponseSchema>;

/**
 * Who owns session state for an agent.
 *
 * - `server`: a session store is configured and snapshots are persisted
 *   server-side.
 * - `client`: no store; state flows through the agent's invocation init
 *   and output payloads.
 */
export const AgentStateManagementSchema = z.enum(['server', 'client']);
export type AgentStateManagement = z.infer<typeof AgentStateManagementSchema>;

/**
 * Zod schema for the agent capability metadata placed under
 * `metadata.agent` on an agent's action descriptor. Lets the Dev UI
 * and other reflective callers render the right surface (e.g. hide
 * the Abort button when the configured store doesn't support it)
 * without round-tripping through the reflection API.
 */
export const AgentMetadataSchema = z.object({
  /** Who owns session state for this agent. */
  stateManagement: AgentStateManagementSchema,
  /**
   * Whether the agent's invocations can be aborted. True only when the
   * configured store implements the abort lifecycle.
   */
  abortable: z.boolean(),
  /**
   * JSON schema for the agent's custom session state (the `custom` field
   * of `SessionState`), inferred from the agent's state type. Lets the
   * Dev UI and other reflective callers render or validate state without
   * the agent describing it separately. Omitted when the state type
   * carries no schema to infer (e.g. an unstructured `any` state).
   */
  stateSchema: z.record(z.any()).optional(),
});
export type AgentMetadata = z.infer<typeof AgentMetadataSchema>;

File: go/genkit/exp/agent.go
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package exp

import (
	"sort"

	aix "github.com/firebase/genkit/go/ai/exp"
	"github.com/firebase/genkit/go/core/api"
	"github.com/firebase/genkit/go/genkit"
	"github.com/firebase/genkit/go/internal/genkitbridge"
)

// DefineAgent defines an agent backed by an inline prompt and registers it as
// an action on the registry. Returns an [aix.Agent].
//
// An Agent is a stateful, multi-turn conversational action. It builds on
// bidirectional streaming to enable ongoing conversations where each turn's
// input and output are streamed between client and server. The framework
// handles session state, conversation history, and optional snapshot
// persistence automatically.
//
// The prompt is defined inline via [aix.InlinePrompt] and registered under the
// agent's name; it accepts every option [genkit.DefinePrompt] does. To back the
// agent with a prompt already in the registry (e.g. one from a .prompt file),
// use [DefinePromptAgent] instead.
//
// Each turn hands the session's conversation to the prompt, so a prompt that
// declares messages of its own places it (see [aix.InlinePrompt]); one that
// declares none has it placed between the system message and the user prompt.
//
// The State type parameter is inferred from the typed agent options
// (e.g. [aix.WithSessionStore], [aix.WithStateTransform]); pass an explicit
// [State] only when no typed option is provided.
//
// The returned agent is an [api.BidiAction]; pass it to [genkit.Handler] to
// serve it over HTTP, one turn per request. Server-managed agents also
// register companion actions for the snapshot lifecycle; serve them
// alongside the agent via [aix.Agent.GetSnapshotAction] and
// [aix.Agent.AbortAction].
//
// For full control over the per-turn loop, use [DefineCustomAgent].
//
// # Options
//
//   - [aix.WithSessionStore]: Enable snapshot persistence
//   - [aix.WithStateTransform]: Rewrite session state on its way out to the client
//   - [aix.WithStreamTransform]: Rewrite stream chunks on their way out to the client
//
// Example:
//
//	chatAgent := exp.DefineAgent(g, "chat",
//		aix.InlinePrompt{
//			ai.WithModelName("googleai/gemini-flash-latest"),
//			ai.WithSystem("You are a helpful assistant."),
//		},
//		aix.WithSessionStore(localstore.NewInMemorySessionStore[any]()),
//	)
func DefineAgent[State any](
	g *genkit.Genkit,
	name string,
	prompt aix.InlinePrompt,
	opts ...aix.AgentOption[State],
) *aix.Agent[State] {
	requireExperimental(g, "DefineAgent")
	return aix.DefineAgent(genkitbridge.RegistryOf(g), name, prompt, opts...)
}

// DefinePromptAgent defines a prompt-backed agent sourced from the registry by
// name and registers it as an action. Returns an [aix.Agent].
//
// By default the agent uses the prompt registered under its own name (e.g. one
// defined via [genkit.DefinePrompt] or loaded from a .prompt file), so no source
// option is required. Pass [aix.WithNamedPrompt] to reference a differently
// named prompt and supply its render input from code, so a single prompt can
// back many agents.
//
// It is the registry-backed counterpart of [DefineAgent]: where [DefineAgent]
// defines the prompt inline, DefinePromptAgent points at a prompt already in
// the registry. The prompt source is a typed option ([aix.WithNamedPrompt])
// rather than a positional argument, so it composes with the other agent
// options in one variadic. For full control over the per-turn loop, use
// [DefineCustomAgent].
//
// Each turn hands the session's conversation to the prompt, so a prompt that
// declares messages of its own places it with {{history}} or
// [ai.HistoryFromContext] (see [aix.InlinePrompt] for the rule); one that
// declares none has it placed between the system message and the user prompt.
// A .prompt file's body is a multi-turn template, so {{history}} works in it
// directly.
//
// The State type parameter is inferred from the typed agent options; pass an
// explicit [State] only when no typed option provides it.
//
// # Options
//
//   - [aix.WithNamedPrompt]: Source from a differently named prompt with a code-supplied input
//   - [aix.WithSessionStore]: Enable snapshot persistence
//   - [aix.WithStateTransform]: Rewrite session state on its way out to the client
//   - [aix.WithStreamTransform]: Rewrite stream chunks on their way out to the client
//
// Example (same-named prompt loaded from ./prompts/chef.prompt):
//
//	chef := exp.DefinePromptAgent(g, "chef",
//		aix.WithSessionStore(localstore.NewInMemorySessionStore[any]()),
//	)
//
// Example (a shared prompt, parameterized per agent):
//
//	pirate := exp.DefinePromptAgent(g, "pirate",
//		aix.WithNamedPrompt[any]("chat", ChatInput{Personality: "a sarcastic pirate"}),
//	)
func DefinePromptAgent[State any](
	g *genkit.Genkit,
	name string,
	opts ...aix.PromptAgentOption[State],
) *aix.Agent[State] {
	requireExperimental(g, "DefinePromptAgent")
	return aix.DefinePromptAgent(genkitbridge.RegistryOf(g), name, opts...)
}

// DefineCustomAgent defines an agent with full control over the conversation
// loop, registers it as an action of type agent, and returns an
// [aix.Agent].
//
// The provided function fn receives a [aix.Responder] for streaming output
// to the client and an [aix.SessionRunner] for accessing conversation state.
// Call [aix.SessionRunner.Run] to enter the turn loop, which blocks until the
// client sends the next message.
//
// Like [DefineAgent], the returned agent is an [api.BidiAction] servable
// via [genkit.Handler], with companion actions on [aix.Agent.GetSnapshotAction]
// and [aix.Agent.AbortAction].
//
// For agents backed by a prompt, use [DefineAgent] (inline prompt) or
// [DefinePromptAgent] (a prompt already in the registry) instead.
//
// # Options
//
//   - [aix.WithSessionStore]: Enable snapshot persistence
//   - [aix.WithStateTransform]: Rewrite session state on its way out to the client
//   - [aix.WithStreamTransform]: Rewrite stream chunks on their way out to the client
//
// The State type parameter is the shape of the conversation's custom state
// ([aix.SessionState.Custom]); mutating it via [aix.Session.UpdateCustom]
// streams an [aix.AgentStreamChunk.CustomPatch] delta to the client.
//
// Example:
//
//	chatAgent := exp.DefineCustomAgent(g, "chat",
//		func(ctx context.Context, resp aix.Responder, sess *aix.SessionRunner[any]) (*aix.AgentResult, error) {
//			var lastMessage *ai.Message
//			err := sess.Run(ctx, func(ctx context.Context, input *aix.AgentInput) (*aix.TurnResult, error) {
//				var reason aix.AgentFinishReason
//				for result, err := range genkit.GenerateStream(ctx, g,
//					ai.WithModelName("googleai/gemini-3-flash-preview"),
//					ai.WithMessages(sess.Messages()...),
//				) {
//					if err != nil {
//						return nil, err
//					}
//					if result.Done {
//						lastMessage = result.Response.Message
//						reason = aix.AgentFinishReason(result.Response.FinishReason)
//						sess.AddMessages(lastMessage)
//					} else {
//						resp.SendModelChunk(result.Chunk)
//					}
//				}
//				// Report how the turn ended; the framework forwards it on
//				// the TurnEnd chunk and persists it on the snapshot.
//				return &aix.TurnResult{FinishReason: reason}, nil
//			})
//			if err != nil {
//				return nil, err
//			}
//			return &aix.AgentResult{Message: lastMessage}, nil
//		},
//	)
func DefineCustomAgent[State any](
	g *genkit.Genkit,
	name string,
	fn aix.AgentFunc[State],
	opts ...aix.AgentOption[State],
) *aix.Agent[State] {
	requireExperimental(g, "DefineCustomAgent")
	return aix.DefineCustomAgent(genkitbridge.RegistryOf(g), name, fn, opts...)
}

// LookupAgent resolves an agent registered on g by name and returns its
// [aix.AgentHandle]: the untyped caller-side view for code that does not hold
// the typed [aix.Agent] value (orchestrators, middleware, tools). The handle
// runs the agent ([aix.AgentHandle.Run], [aix.AgentHandle.RunText]), launches
// and tracks background work ([aix.AgentHandle.RunDetached],
// [aix.AgentHandle.Task]), and reads or aborts snapshots, with custom
// state as raw JSON.
//
// It returns nil when name resolves to no agent, matching every other Lookup
// in the framework. Like [ListAgents] it does not require
// [genkit.WithExperimental]: it only reads the registry, and an agent can only
// have been registered through the gated constructors.
func LookupAgent(g *genkit.Genkit, name string) *aix.AgentHandle {
	return aix.LookupAgent(genkitbridge.RegistryOf(g), name)
}

// ListAgents returns a slice of all [api.Action] instances that represent
// agents registered with the Genkit instance g. Like [genkit.ListFlows], this
// is useful for introspection or for dynamically exposing agent endpoints in an
// HTTP server; an agent served via [genkit.Handler] runs one turn per request.
func ListAgents(g *genkit.Genkit) []api.Action {
	reg := genkitbridge.RegistryOf(g)
	agents := []api.Action{}
	for _, act := range reg.ListActions() {
		if act.Desc().Type == api.ActionTypeAgent {
			agents = append(agents, act)
		}
	}
	sort.Slice(agents, func(i, j int) bool {
		return agents[i].Name() < agents[j].Name()
	})
	return agents
}
